core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -171,6 +171,23 @@
   private volatile MapIterator destructiveIterator = null;
   private LinkedList<UnsafeSorterSpillWriter> spillWriters = new LinkedList<>();
 
+  private boolean anyNullKeyExists = false;
+
+  public boolean inputEmpty()
+  {
+    return ((numKeys == 0) && !anyNullKeyExists);
+  }
+
+  public boolean isAnyNullKeyExists()
+  {
+    return anyNullKeyExists;
+  }
+
+  public void setAnyNullKeyExists(boolean anyNullKeyExists)
+  {
+    this.anyNullKeyExists = anyNullKeyExists;
```

Comment:
So just making sure I am reading this code right: There is no extra scan of the rows done to know if there are no null keys.

Reply:
```
I set anyNullKeyExists during go through the input iterator

while (input.hasNext) {
      val row = input.next().asInstanceOf[UnsafeRow]
      numFields = row.numFields()
      val key = keyGenerator(row)
      if (!key.anyNull) {
        val loc = binaryMap.lookup(key.getBaseObject, key.getBaseOffset, key.getSizeInBytes)
        val success = loc.append(
          key.getBaseObject, key.getBaseOffset, key.getSizeInBytes,
          row.getBaseObject, row.getBaseOffset, row.getSizeInBytes)
        if (!success) {
          binaryMap.free()
          // scalastyle:off throwerror
          throw new SparkOutOfMemoryError("There is not enough memory to build hash map")
          // scalastyle:on throwerror
        }
      } else {
        binaryMap.setAnyNullKeyExists(true)   // HERE
      }
    }



```